alibi_detect/saving/validators.py

```diff
@@ -0,0 +1,84 @@
+import sys
+from typing import Any, Generic, Optional, Type, TypeVar, Union, List
+import numpy as np
+from numpy.lib import NumpyVersion
+from pydantic.fields import ModelField
+
+from alibi_detect.utils.frameworks import has_tensorflow, has_pytorch, has_keops, Framework
+
+if has_tensorflow:
+    import tensorflow as tf
+if has_pytorch:
+    import torch
+
+
+def coerce_int2list(value: int) -> List[int]:
+    """Validator to coerce int to list (pydantic doesn't do this by default)."""
+    if not isinstance(value, list):
+        return [value]
+    else:
+        return value
+
+
+# Framework validator (validates `flavour` and `backend` fields)
+def validate_framework(framework: str, field: ModelField) -> str:
+    if (framework == Framework.TENSORFLOW and has_tensorflow) or (framework == Framework.PYTORCH and has_pytorch) or \
+            (framework == Framework.KEOPS and has_keops):
+        return framework
+    elif framework == Framework.SKLEARN:  # sklearn is a core dep
+        return framework
+    else:
+        raise ImportError(f"`{field.name} = '{framework}'` not possible since {framework} is not installed.")
+
+
+# NumPy NDArray pydantic validator type
+# The code below is adapted from https://github.com/cheind/pydantic-numpy.
+T = TypeVar("T", bound=np.generic)
+if NumpyVersion(np.__version__) < "1.22.0" or sys.version_info < (3, 9):
+    class NDArray(Generic[T], np.ndarray):
+        """
+        A Generic pydantic model to coerce to np.ndarray's.
+        """
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, val: Any, field: ModelField) -> np.ndarray:
+            return _coerce2NDArray(cls, val, field)
+
+else:
+    class NDArray(Generic[T], np.ndarray[Any, T]):  # type: ignore[no-redef, type-var]
+        """
+        A Generic pydantic model to coerce to np.ndarray's.
+        """
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, val: Any, field: ModelField) -> Optional[np.ndarray]:
+            return _coerce2NDArray(cls, val, field)
+
+
+def _coerce2NDArray(cls: Type, val: Any, field: ModelField) -> np.ndarray:
+    if field.sub_fields is not None:
+        dtype_field = field.sub_fields[0]
+        return np.asarray(val, dtype=dtype_field.type_)
+    else:
+        return np.asarray(val)
+
+
+def coerce_2tensor(value: Union[float, List[float]], values: dict):
```

Comment:
Type information ambiguous, e.g. no return type and also seems `None` is an option?

Reply:
Thinking about this more, I can add a unit test for `coerce_2tensor` to `test_validate`. But for `validate_framework` I'm not so sure what the best approach is i.e. how to mock `has_*`. @mauicv any ideas?